File: src/main/java/com/yupi/airouter/service/ApiKeyService.java
package com.yupi.airouter.service;

import com.mybatisflex.core.service.IService;
import com.yupi.airouter.model.entity.ApiKey;
import com.yupi.airouter.model.entity.User;

import java.util.List;

/**
 * API Key 服务
 *
 * @author <a href="https://codefather.cn">编程导航学习圈</a>
 */
public interface ApiKeyService extends IService<ApiKey> {

    /**
     * 创建 API Key
     *
     * @param keyName Key名称/备注
     * @param loginUser 当前登录用户
     * @return API Key
     */
    ApiKey createApiKey(String keyName, User loginUser);

    /**
     * 获取用户的 API Key 列表
     *
     * @param userId 用户ID
     * @return API Key 列表
     */
    List<ApiKey> listUserApiKeys(Long userId);

    /**
     * 撤销 API Key
     *
     * @param id API Key ID
     * @param userId 用户ID
     * @return 是否成功
     */
    boolean revokeApiKey(Long id, Long userId);

    /**
     * 根据 Key 值查询 API Key
     *
     * @param keyValue Key值
     * @return API Key
     */
    ApiKey getByKeyValue(String keyValue);

    /**
     * 更新 API Key 的使用统计
     *
     * @param apiKeyId API Key ID
     * @param tokens 本次消耗的Token数
     */
    void updateUsageStats(Long apiKeyId, Integer tokens);
}

File: go-backend/internal/service/apikey_service.go
package service

import (
	"crypto/rand"
	"encoding/hex"
	"strings"

	"github.com/yupi/airouter/go-backend/internal/common"
	"github.com/yupi/airouter/go-backend/internal/errno"
	"github.com/yupi/airouter/go-backend/internal/model/entity"
	"github.com/yupi/airouter/go-backend/internal/model/vo"
	"github.com/yupi/airouter/go-backend/internal/repository"
)

const (
	apiKeyPrefix             = "sk-"
	apiKeyRandomHexBytes     = 16
	apiKeyMaskMinLength      = 12
	apiKeyMaskPrefixLength   = 8
	apiKeyMaskSuffixLength   = 4
)

type ApiKeyService struct {
	apiKeyRepo *repository.ApiKeyRepository
}

func NewApiKeyService(apiKeyRepo *repository.ApiKeyRepository) *ApiKeyService {
	return &ApiKeyService{apiKeyRepo: apiKeyRepo}
}

func (s *ApiKeyService) CreateApiKey(keyName string, loginUser *entity.User) (*entity.ApiKey, error) {
	keyValue, err := generateAPIKey()
	if err != nil {
		return nil, errno.New(errno.SystemError)
	}
	apiKey := &entity.ApiKey{
		UserID:      loginUser.ID,
		KeyValue:    keyValue,
		KeyName:     keyName,
		Status:      "active",
		TotalTokens: 0,
	}
	if err = s.apiKeyRepo.Create(apiKey); err != nil {
		return nil, errno.New(errno.SystemError)
	}
	return apiKey, nil
}

func (s *ApiKeyService) ListMyApiKeys(userID, pageNum, pageSize int64) (common.PageResponse[vo.ApiKeyVO], error) {
	list, total, err := s.apiKeyRepo.ListByUser(userID, pageNum, pageSize)
	if err != nil {
		return common.PageResponse[vo.ApiKeyVO]{}, errno.New(errno.SystemError)
	}
	voList := make([]vo.ApiKeyVO, 0, len(list))
	for _, item := range list {
		voList = append(voList, vo.ApiKeyVO{
			ID:           item.ID,
			KeyValue:     maskAPIKey(item.KeyValue),
			KeyName:      item.KeyName,
			Status:       item.Status,
			TotalTokens:  item.TotalTokens,
			LastUsedTime: item.LastUsedTime,
			CreateTime:   item.CreateTime,
		})
	}
	return common.BuildPageResponse(voList, pageNum, pageSize, total), nil
}

func (s *ApiKeyService) RevokeApiKey(id, userID int64) (bool, error) {
	ok, err := s.apiKeyRepo.RevokeByIDAndUser(id, userID)
	if err != nil {
		return false, errno.New(errno.SystemError)
	}
	if !ok {
		return false, errno.NewWithMessage(errno.NotFoundError, "API Key 不存在")
	}
	return true, nil
}

func (s *ApiKeyService) GetByID(id int64) (*entity.ApiKey, error) {
	apiKey, err := s.apiKeyRepo.GetByID(id)
	if err != nil {
		return nil, errno.New(errno.SystemError)
	}
	return apiKey, nil
}

func (s *ApiKeyService) GetByKeyValue(keyValue string) (*entity.ApiKey, error) {
	apiKey, err := s.apiKeyRepo.GetByKeyValueActive(keyValue)
	if err != nil {
		return nil, errno.New(errno.SystemError)
	}
	return apiKey, nil
}

func (s *ApiKeyService) UpdateUsageStats(apiKeyID int64, tokens int) {
	_ = s.apiKeyRepo.AddUsageStats(apiKeyID, tokens)
}

func generateAPIKey() (string, error) {
	randomBytes := make([]byte, apiKeyRandomHexBytes)
	if _, err := rand.Read(randomBytes); err != nil {
		return "", err
	}
	return apiKeyPrefix + hex.EncodeToString(randomBytes), nil
}

func maskAPIKey(key string) string {
	if len(key) <= apiKeyMaskMinLength {
		return key
	}
	prefix := key[:apiKeyMaskPrefixLength]
	suffix := key[len(key)-apiKeyMaskSuffixLength:]
	return strings.Join([]string{prefix, "****", suffix}, "")
}
